src/imars3dv2/filters/crop.py

```diff
@@ -0,0 +1,162 @@
+        The image stack to crop. Can also be a 2D image.
+    @param crop_limit: tuple
+        The four limits for cropping. Default is (-1, -1, -1, -1), which will trigger
+        the automatic bounds detection.
+    @param border_pix: int
+        the width of border region to estimate the background intensity, which helps
+        to determine which case we are in.
+    @param expand_ratio: float
+        the ratio to expand the cropped region.
+    @param rel_intensity_threshold_air_or_slit: float
+        passing through keyword arguments to detect_bounds.
+    @param rel_intensity_threshold_fov: float
+        passing through keyword arguments to detect_bounds.
+    @param rel_intensity_threshold_sample: float
+        passing through keyword arguments to detect_bounds.
+
+    Returns
+    -------
+    @return: np.ndarray
+        The cropped image stack.
+    """
+    # check if crop limits (bounding box) are provided
+    if -1 in crop_limit:
+        crop_limit = detect_bounds(
+            arrays,
+            border_pix,
+            expand_ratio,
+            rel_intensity_threshold_air_or_slit,
+            rel_intensity_threshold_fov,
+            rel_intensity_threshold_sample,
+        )
+    # crop
+    left, right, top, bottom = crop_limit
+    if arrays.ndim == 2:
+        return arrays[top:bottom, left:right]
+    elif arrays.ndim == 3:
+        return arrays[:, top:bottom, left:right]
+    else:
+        raise ValueError("Only 2D and 3D arrays are supported.")
+
+
+def detect_bounds(
+    arrays: np.ndarray,
+    border_pix: int = 10,
+    expand_ratio: float = 0.1,
+    rel_intensity_threshold_air_or_slit: float = 0.05,
+    rel_intensity_threshold_fov: float = 0.1,
+    rel_intensity_threshold_sample: float = 0.95,
+) -> tuple:
+    """
+    Auto detect bounds based on intensity thresholding.
+
+    @param arrays: np.ndarray
+        The image stack to crop. Can also be a 2D image.
+    @param border_pix: int
+        the width of border region to estimate the background intensity
+    @param expand_ratio: float
+        the ratio to expand the cropped region.
+    @param rel_intensity_threshold_air_or_slit: float
+        the relative intensity threshold to determine whether the outter boarder
+        is slit (case 1) or air (case 2).
+    @param rel_intensity_threshold_fov: float
+        the relative intensity threshold used to determine pixels within to the
+        field of view, only valid for case 1.
+    @param rel_intensity_threshold_sample: float
+        the relative intensity threshold used to determine pixels within the
+        sample region, only valid for case 2, and the value is relative to the
+        intensity of the air region (outter region in case 2).
+
+    @return: tuple
+        The crop limits in (left, right, top, bottom) order.
+    """
+    # generate representative image
+    if arrays.ndim == 2:
+        img = arrays
+    elif arrays.ndim == 3:
+        img = arrays.mean(axis=0)
```

Comment:
Doesn't this mean you have to have the whole stack available at once? Should we have a way of annotating that requirement that we can search in the code?

Reply:
For large data set that cannot be hold in memory, we can use rolling sum to create a 2D image outside this function, then pass the average image (2D) to extract the bound.